### PR TITLE
PE-379 

### PR DESCRIPTION
**Pull request by andrey-canon**
## Description 
- Allow upload of Template HTML and CSS ([PE-386](https://proversity.atlassian.net/browse/PE-386))
- Update Recap to look for and use custom_report_format ([PE-389](https://proversity.atlassian.net/browse/PE-389))

### CODE REVIEW

File: recap/recap.py [hunk 1]
```diff
@@ -162,28 +162,18 @@ def get_blocks_filtering_list(self, xblock_list):
...
-    def get_submission_key(self, usage_key):
+    def get_submission_key(self, usage_key, user=None):
         """
         Returns submission key needed for submissions api
         """
         try:
             logger.info('Attempting to retrieve student item dictionary.')
-            user = self.runtime.get_real_user(self.runtime.anonymous_student_id)
+            if not user:
+                user = self.runtime.get_real_user(self.runtime.anonymous_student_id)
```
Comment (Squirrel18) [inline recap/recap.py]: Please, add a line break here. @andrey-canon

File: recap/recap.py [hunk 2]
```diff
@@ -225,165 +213,44 @@ def get_display_answer(self, answer):
...
-
     @XBlock.supports("multi_device")
     def student_view(self, context=None):
         """
         The primary view of the RecapXBlock seen in LMS
         """
         blocks = []
         for usage_key, xblock_type in self.get_blocks(self.xblock_list):
-            if xblock_type == 'freetextresponse':
+
```
Comment (Squirrel18) [inline recap/recap.py]: No line break required here. @andrey-canon

File: recap/recap.py [hunk 3]
```diff
@@ -225,165 +213,44 @@ def get_display_answer(self, answer):
...
+            block = self.runtime.get_block(usage_key)
```
Comment (Squirrel18) [inline recap/recap.py]: Please add a line break here. @andrey-canon

File: recap/recap.py [hunk 4]
```diff
@@ -225,165 +213,44 @@ def get_display_answer(self, answer):
...
+            if hasattr(block, 'custom_report_format'):
+                question = unicode(block.display_name)
```
Comment (Squirrel18) [inline recap/recap.py]: Please, add a aline break here. @andrey-canon

File: recap/recap.py [hunk 5]
```diff
@@ -225,165 +213,44 @@ def get_display_answer(self, answer):
...
-                        answer = self.get_answer(usage_key, block, answer_field)
-                        blocks.append((getattr(block, question_field), answer))
-                except Exception as e:
-                    logger.warn(str(e))
-                    logger.info(
-                        'The submissions api failed, using default module store.'
-                    )
+                    user = self.runtime.get_real_user(self.runtime.anonymous_student_id)
+                except TypeError:
+                    user = None
```
Comment (Squirrel18) [inline recap/recap.py]: Please, add a aline break here.

File: recap/recap.py [hunk 6]
```diff
@@ -225,165 +213,44 @@ def get_display_answer(self, answer):
...
+                answer = block.custom_report_format(
+                    user=user,
+                    block=block,
+                )
```
Comment (Squirrel18) [inline recap/recap.py]: Please, add a aline break here. @andrey-canon

File: recap/recap.py [hunk 7]
```diff
@@ -435,84 +305,88 @@ def student_view(self, context=None):
...
-                        block = self.runtime.get_block(usage_key)
-                        question_field, answer_field = self.get_field_names(xblock_type)
-                        answer = self.get_user_answer(usage_key, block, answer_field, user)
-                        blocks.append((getattr(block, question_field), answer))
-                    except Exception as e:
-                        logger.warn(str(e))
-                        logger.info(
-                            'The submissions api failed, using default module store.'
-                        )
+                block = self.runtime.get_block(usage_key)
```
Comment (Squirrel18) [inline recap/recap.py]: Please, add a aline break here. @andrey-canon

File: recap/recap.py [hunk 8]
```diff
@@ -435,84 +305,88 @@ def student_view(self, context=None):
...
+                if hasattr(block, 'custom_report_format'):
+                    question = unicode(block.display_name)
+                    answer = block.custom_report_format(
+                        user=user,
+                        block=block,
+                    )
```
Comment (Squirrel18) [inline recap/recap.py]: Please, add a aline break here. @andrey-canon

File: recap/recap.py [hunk 9]
```diff
@@ -435,84 +305,88 @@ def student_view(self, context=None):
...
-        layout = ''
+
+        def update_layout(layout, pattern, use_block_layout=False):
+            '''
+            '''
+            block_sets = []
+            for m in re.finditer(pattern, layout):
+                try:
+                    title = blocks[int(m.group(1)) - 1][0]
+                    answer = blocks[int(m.group(1)) - 1][1]
```
Comment (Squirrel18) [inline recap/recap.py]: Add a line break. @andrey-canon

File: recap/recap.py [hunk 10]
```diff
@@ -435,84 +305,88 @@ def student_view(self, context=None):
...
+                        qa_str = unicode(block_layout).format(title, self.get_display_answer(answer))
+                    elif m.group(2) == 'TITLE':
+                        qa_str = title
+                    elif m.group(2) == 'ANSWER':
+                        qa_str = unicode(self.get_display_answer(answer))
+                    else:
+                        qa_str = _('Bad Format')
+                    block_sets.append((m.start(0), m.end(0), qa_str))
+                except IndexError:
+                    pass
```
Comment (Squirrel18) [inline recap/recap.py]: Maybe, it could be useful to log some kind of message. @andrey-canon

File: recap/recap.py [hunk 11]
```diff
@@ -435,84 +305,88 @@ def student_view(self, context=None):
...
+
+            for start, end, string in reversed(block_sets):
```
Comment (Squirrel18) [inline recap/recap.py]: Why reversed is required here? @andrey-canon

File: recap/recap.py [hunk 12]
```diff
@@ -435,84 +305,88 @@ def student_view(self, context=None):
...
+                layout = layout[0:start] + string + layout[end:]
+
+            return layout
+
+        if self.html_file:
```
Comment (Squirrel18) [inline recap/recap.py]: This could be a inline if assignment. @andrey-canon

File: recap/recap.py [hunk 13]
```diff
@@ -637,44 +509,42 @@ def make_pdf_json(self, data, suffix=''):
...
         downloadable_users = User.objects.filter(
             courseenrollment__course_id=course_id,
             courseenrollment__is_active=1,
             id__in=map(int, student_ids_intersection)
         ).values('username', 'email', 'id')
 
-        return  {"data": list(downloadable_users) }
-        
-
+        return {"data": list(downloadable_users)}
```
Comment (Squirrel18) [inline recap/recap.py]: Please, don't mix single quotes and double quotes. @andrey-canon

File: recap/static/js/src/recap_edit.js [hunk 14]
```diff
@@ -125,6 +131,8 @@ function StudioEditableXBlockMixin(runtime, element) {
 
         // Count all the number of items in the Xblock list
         var counter = 0;
+        var inputHtmlFile= $('#xb-field-edit-html_file');
```
Comment (Squirrel18) [inline recap/static/js/src/recap_edit.js]: Please add a black sapce between the variable name and equal char. @andrey-canon

File: recap/static/js/src/recap_edit.js [hunk 15]
```diff
@@ -125,6 +131,8 @@ function StudioEditableXBlockMixin(runtime, element) {
...
+        var inputCssFile= $('#xb-field-edit-css_file');
```
Comment (Squirrel18) [inline recap/static/js/src/recap_edit.js]: Same previous comment. @andrey-canon

File: recap/static/js/src/recap_edit.js [hunk 16]
```diff
@@ -157,9 +166,17 @@ function StudioEditableXBlockMixin(runtime, element) {
             newTextBoxDiv.appendTo("#TextBoxesGroup");
             counter++;
         });
+        htmlResult = $('#xb-field-download-html_file').attr("value");
```
Comment (Squirrel18) [inline recap/static/js/src/recap_edit.js]: Please, add a line break. @andrey-canon

File: recap/static/js/src/recap_edit.js [hunk 17]
```diff
@@ -157,9 +166,17 @@ function StudioEditableXBlockMixin(runtime, element) {
...
+        if (htmlResult){
```
Comment (Squirrel18) [inline recap/static/js/src/recap_edit.js]: Please, add a blank space before open curly braces. @andrey-canon

File: recap/static/js/src/recap_edit.js [hunk 18]
```diff
@@ -157,9 +166,17 @@ function StudioEditableXBlockMixin(runtime, element) {
...
+            var textAreaParent = $('#xb-field-edit-string_html').parent();
+            textAreaParent.css('pointer-events', 'none');
+            textAreaParent.css('opacity', '0.4');
+        }
+        cssResult = $('#xb-field-download-css_file').attr("value");
```
Comment (Squirrel18) [inline recap/static/js/src/recap_edit.js]: Please, don't mix single quotes and double quotes. @andrey-canon

File: recap/static/js/src/recap_edit.js [hunk 19]
```diff
@@ -221,4 +241,24 @@ function StudioEditableXBlockMixin(runtime, element) {
         e.preventDefault();
         runtime.notify('cancel', {});
     });
-}
\ No newline at end of file
+
+    function readInputFile(input){
```
Comment (Squirrel18) [inline recap/static/js/src/recap_edit.js]: Please, add a blank space before open curly braces. @andrey-canon

File: recap/static/js/src/recap_edit.js [hunk 20]
```diff
@@ -221,4 +241,24 @@ function StudioEditableXBlockMixin(runtime, element) {
...
+        var file = input.target.files[0];
+        var targetId = input.target.id;
+        var reader = new FileReader();
+        if (file) {
+            reader.onload = function(e) {
+                if (targetId == 'xb-field-edit-html_file'){
```
Comment (Squirrel18) [inline recap/static/js/src/recap_edit.js]: Please, add a blank space before open curly braces. @andrey-canon

File: recap/static/js/src/recap_edit.js [hunk 21]
```diff
@@ -221,4 +241,24 @@ function StudioEditableXBlockMixin(runtime, element) {
...
+                    var textAreaParent = $('#xb-field-edit-string_html').parent();
+                    textAreaParent.css('pointer-events', 'none');
+                    textAreaParent.css('opacity', '0.4')
+                } else {
+                    cssResult = e.target.result;
+                }
+            };
+            reader.readAsText(file);
+        }
+
```
Comment (Squirrel18) [inline recap/static/js/src/recap_edit.js]: Not line break required here. @andrey-canon

File: recap/recap.py (hunk 13 above)
Comment (andrey-canon) [reply recap/recap.py]: changed

**Pull request merged**